public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx

```diff
@@ -154,6 +133,13 @@ export const QueryAndExpressionsStep = ({ editingExistingRule, onDataChange }: P
 
   const onChangeQueries = useCallback(
     (updatedQueries: AlertQuery[]) => {
+      // Most data sources triggers onChange and onRunQueries consecutively
+      // It means our reducer state is always one step behind when runQueries is invoked
+      // Invocation cycle => onChange -> dispatch(setDataQueries) -> onRunQueries -> setDataQueries Reducer
+      // No matter if we use query or getValues('queries') their state is always stale when we invoke runQueries
+      // As a workaround we use this ref which is updated immediately in onChange and used in runQueries
+      setValue('queries', updatedQueries, { shouldValidate: false });
```

Comment:
I'm not sure if the explanation above matches the code here; it looks like we're updating the form values instead of mutating a ref – or is this a reference to `setValue` mutating a value synchronously?

Reply:
You're correct. I changed the code, but I haven't updated the explanation. I'll update